tests/Properties.hs

```diff
@@ -239,12 +239,20 @@ prop_foldrBP     = eq3
     (L.foldr     :: (W -> X -> X) -> X -> B -> X)
     (P.foldr     :: (W -> X -> X) -> X -> P -> X)
 prop_foldrBP'    = eq3
-    (L.foldr     :: (W -> X -> X) -> X -> B -> X)
+    (L.foldr'    :: (W -> X -> X) -> X -> B -> X)
     (P.foldr'    :: (W -> X -> X) -> X -> P -> X)
 prop_mapAccumLBP = eq3
     (L.mapAccumL :: (X -> W -> (X,W)) -> X -> B -> (X, B))
     (P.mapAccumL :: (X -> W -> (X,W)) -> X -> P -> (X, P))
 
+explosiveTail :: B -> B
+explosiveTail = (`L.append` error "Tail of this byte string is undefined!")
+
+prop_Lfoldr_lazy = \ xs -> genericTake (L.length xs) (L.foldr (:) [ ] (explosiveTail xs)) == L.unpack xs
+prop_Lfoldr'_strict = expectFailure $ \ xs ys -> genericTake (L.length xs) (L.foldr' (:) [ ] (explosiveTail (xs `L.append` ys))) == L.unpack xs
+prop_Lfoldr1_lazy = \ xs -> L.length xs > 0 ==> L.foldr1 const (explosiveTail (xs `L.append` L.singleton 1)) == L.head xs
+prop_Lfoldr1'_strict = expectFailure $ \ xs ys -> L.length xs > 0 ==> L.foldr1' const (explosiveTail (xs `L.append` L.singleton 1 `L.append` ys)) == L.head xs
```

Comment:
Very nice!

Reply:
Glad you like it. It took a while to get these right!